Run all tests in multi-nodes deployment in CI

gdemonet opened the pull request:
## What

This PR introduces a new VM in the Terraform multi-nodes spawn process, called _**bastion**_. It allows one to have access to the private networks spawned, unlike the Eve worker.

The post-install tests (defined under `tests/post`) are now all run without custom options in the multi-nodes deployment context (previously, we invoked `pytest` with custom CLI options `--bootstrap-ip` and `--skip-tls-verify` for performing KubeAPI calls from outside the control-plane network).

## How to test

Look at Eve build reports, see that the tests all run in the multi-nodes stage (no test is skipped).

## Left to do

Some tasks, related to this PR, will need to be tackled in (yet-to-be-created) other issues:

- One would want the bastion to run the install scenario instead of the worker
- One would want to have a more friendly Terraform tooling, usable locally
- One would like to remove `kubectl` invocations in tests, and only use the Python Kubernetes client instead

gdemonet committed:
Create control and workload networks in Terraform

Previously, we only defined a single "private" network for the spawned
nodes in Terraform. We now define one for control-plane and one for
workload-plane, and attach all nodes to them.

This change impacts the `BootstrapConfiguration` we shipped with this
Terraform deployment.

Issue: GH-889

gdemonet committed:
Add bastion VM as part of multi-nodes deployment

We previously considered the OpenStack worker from Eve as the bastion
from which to orchestrate multi-nodes deployments in the CI. However,
since we cannot attach this worker to the private networks it deployed,
we introduce another VM, which we call "bastion", that will assume (in
the long-term) the responsibility of installing and testing the product.

Issue: GH-889

gdemonet committed:
Generate SSH config files from Terraform

These files (one for the worker, on for the bastion) need data from
the Terraform state, so we decided to generate them using Terraform
templating.

Issue: GH-889

gdemonet committed:
Generate and share bastion SSH keypair

Issue: GH-889

gdemonet committed:
Run test suite from Bastion

This approach naively copies test files from the worker to the bastion,
which we may not want to do in the long run. In the meantime, this
allows tests to work the same way in single and multi-nodes deployments.

Issue: GH-889

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/889-multi-nodes-bastion` into
`development/2.0`
* [w/2.1/improvement/889-multi-nodes-bastion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/889-multi-nodes-bastion) will be merged into `development/2.1` 
* [w/2.2/improvement/889-multi-nodes-bastion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/889-multi-nodes-bastion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented on eve/workers/openstack-multiple-nodes/terraform/network.tf:
```diff
@@ -6,9 +8,62 @@ variable "openstack_network" {
+# Default internal networks
+variable "control_plane_cidr" {
+  type    = "string"
+  default = "172.21.254.0/28"
```
Can I suggest to use *another* subnet for these two, such that we never accidentally make assumptions on the Vagrant environment being 'the default'?

gdemonet replied:
Yes we can, what would fit in your opinion? `172.22.254.X` instead of `172.21.254.X` would be OK?

gdemonet replied:
Done: 070c4a8

gdemonet committed:
Use different CIDR for Terraform and Vagrant

To guard us against leaving hard-coded values, we should make sure to
keep the Terraform and Vagrant environments different in as many
constants as possible (network ranges, hostnames, mountpoints...).

Issue: GH-889

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/889-multi-nodes-bastion` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/889-multi-nodes-bastion`).

```bash
 $ git fetch
 $ git checkout origin/improvement/889-multi-nodes-bastion
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/889-multi-nodes-bastion
```

bert-e commented:
# History mismatch



Merge commit #bfd3809ebd7d0fb383ecb854e5d76a64cac11690 on the integration branch
`w/2.1/improvement/889-multi-nodes-bastion` is merging a branch which is neither the current
branch `improvement/889-multi-nodes-bastion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/889-multi-nodes-bastion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/889-multi-nodes-bastion` into
`development/2.0`
* [w/2.1/improvement/889-multi-nodes-bastion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/889-multi-nodes-bastion) will be merged into `development/2.1` 
* [w/2.2/improvement/889-multi-nodes-bastion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/889-multi-nodes-bastion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

gdemonet committed:
Clean script-based provisioning in Terraform

We also extract the few dhclient calls to a script and attempt to retry
until an IP is found (since we had flaky situations where an IP could
be missing).

Issue: GH-889

bert-e commented:
# History mismatch



Merge commit #c19609ef74e3252dfbcdae78421d1a80526514a1 on the integration branch
`w/2.1/improvement/889-multi-nodes-bastion` is merging a branch which is neither the current
branch `improvement/889-multi-nodes-bastion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/889-multi-nodes-bastion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/889-multi-nodes-bastion` into
`development/2.0`
* [w/2.1/improvement/889-multi-nodes-bastion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/889-multi-nodes-bastion) will be merged into `development/2.1` 
* [w/2.2/improvement/889-multi-nodes-bastion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/889-multi-nodes-bastion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented:
Does this actually enable all tests we're currently skipping in multi-node CI?

gdemonet commented:
> Does this actually enable all tests we're currently skipping in multi-node CI?

Yes, because we were using the `--bootstrap-ip` pytest CLI option to "know" whether the tests run in a multi-node context. However, I forgot to remove that code from the tests themselves, I guess I should do it in this PR.

NicolasT commented:
> > Does this actually enable all tests we're currently skipping in multi-node CI?
> 
> Yes, because we were using the `--bootstrap-ip` pytest CLI option to "know" whether the tests run in a multi-node context. However, I forgot to remove that code from the tests themselves, I guess I should do it in this PR.

Yes please :)

gdemonet committed:
Remove unneeded Pytest CLI options

The options "--bootstrap-ip" and "--skip-tls-verify" were only
introduced to cope with the limitations of our multi-nodes deployment
in CI. It should now be fixed.

Issue: GH-889

gdemonet commented:
> Yes please :)

Done: e528597

gdemonet commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
